src/pykeen/nn/node_piece.py

```diff
@@ -685,6 +673,126 @@ def __call__(
         return len(anchors) + 1, torch.as_tensor(tokens, dtype=torch.long)
 
 
+def _random_sample_no_replacement(
+    pool: Mapping[int, Collection[int]],
+    num_tokens: int,
+) -> torch.LongTensor:
+    # randomly sample without replacement num_tokens relations for each entity
+    assignment = torch.full(
+        size=(len(pool), num_tokens),
+        dtype=torch.long,
+        fill_value=-1,
+    )
+    # TODO: vectorization?
+    for idx, this_pool in tqdm(pool.items(), desc="sampling", leave=False, unit_scale=True):
+        this_pool_t = torch.as_tensor(data=list(this_pool), dtype=torch.long)
+        this_pool = this_pool_t[torch.randperm(this_pool_t.shape[0])[:num_tokens]]
+        assignment[idx, : len(this_pool_t)] = this_pool
+    return assignment
+
+
+class PrecomputedTokenizerLoader:
+    """A loader for precomputed tokenization."""
+
+    @abstractmethod
+    def __call__(self, path: pathlib.Path) -> Tuple[Mapping[int, Collection[int]], int]:
+        """Load tokenization from the given path."""
+        raise NotImplementedError
+
+
+class GalkinPickleLoader(PrecomputedTokenizerLoader):
+    """A loader for pickle files provided by Galkin et al."""
+
+    def __call__(self, path: pathlib.Path) -> Tuple[Mapping[int, Collection[int]], int]:  # noqa: D102
+        with path.open(mode="rb") as pickle_file:
+            # contains: anchor_ids, entity_ids, mapping {entity_id -> {"ancs": anchors, "dists": distances}}
+            anchor_ids, mapping = pickle.load(pickle_file)[0::2]
+        logger.info(f"Loaded precomputed pools with {len(anchor_ids)} anchors, and {len(mapping)} pools.")
+        # normalize anchor_ids
+        anchor_map = {a: i for i, a in enumerate(anchor_ids) if a >= 0}
+        # map padding to padding
+        anchor_map[-1] = -1
```

Comment:
given the above comment, the padding map should use -99
```
anchor_map[-99] = -1
```

Reply:
Okay, later on, i'll pad to a fixed length with -1 anyway
https://github.com/pykeen/pykeen/blob/527cb6aa42eb401e891357d1079b2d32c3b9c596/src/pykeen/nn/node_piece.py#L680-L691

Reply:
317ef79